### ISSUE TITLE
Stop shipping dependencies

### ISSUE THREAD

**Issue by sztomi** [rpclib/rpclib#155]
So I've been quietly juggling #152 and failing to resolve a linker error after renaming the namespaces. Since #135 seems fairly popular, I'm more and more inclined to not ship most dependencies, at least not asio and msgpack. The advantages are:

  * More flexibility to end users (and #135 becomes straightforward)
  * Avoiding code duplication when non-rpclib copies of the same libraries are used
  * Way smaller codebase for rpclib

The disadvantage is that compiling the project will require installing at least asio and msgpack. This is not ideal, but can be amended with conan packages of said libraries (#87 is long overdue, too).

cc @sytelus @nylan

**Comment by sytelus** [comment]
I would strongly vote for not doing this. We use rpclib as submodule and our scripts builds it from source on OSX, Linux, Windows. If there are any additional steps that needs to take to compile rpclib is a big issue for us. We had like to keep our build simple and our dependency to minimum (basically just two: Eigen and rpclib). I think dependency-free nature of rpclib is one of its strongest selling point compared to other solutions out there.

**Comment by sztomi** [comment]
@sytelus I hear that, and it was originally done for that exact reason. However, this abstraction is leaky and causes an active disadvantage for some users. Now, if you are building from source as part of your build you don't need to worry about this change, because I'll implement it so that fetching the dependencies is performed from the build script (maybe cmake itself). As an end user wanting the convenience of not having to install the dependencies your life won't change much. At the same time, this gives the option for users already using asio or other dependencies of rpclib to avoid duplication. This is also very important for enterprise scenarios where most users will want an easy way to manage versions of their libraries. Is it going to work for you if this is done in a transparent fashion?

**Comment by sytelus** [comment]
One big issue right now I'm facing is that I need to build rpclib as part of the project. As it is, I'm not able to do this yet successfully and this is preventing us from generating and distributing Linux binaries which had been asked by probably 1 out of every 3 users (because Unreal builds are complex and takes long time). I'm not sure how above will impact this use case.

Related issue which we would be good to re-open: https://github.com/rpclib/rpclib/issues/139

**Comment by ahoarau** [comment]
From my naive point of view, can't we have the best of both worlds ?
Something like : 
```cmake
find_package(Boost QUIET COMPONENTS asio)
if(NOT Boost_ASIO_FOUND)
     # Use the internal Boost asio 
     include_directories(external/asio/include)
     add_library( asio external/asio/asio.cpp)
     set(Boost_LIBRARIES asio)
endif()

# Link against ${Boost_LIBRARIES}
```

**Comment by sztomi** [comment]
Yes, I absolutely think that having the best of both worlds is possible. So @sytelus this is not implemented yet and shouldn't affect you at all. When I make progress I'll link a branch here and ask you to look at it to see if it's causing any problems.

**Comment by chronos38** [comment]
This change would be appreciated from side, either make dependencies optional, addable from a different source or at least change the include guards of visible header includes. I guess the last point this only applies to msgpack.
